src/main/java/net/earthcomputer/clientcommands/features/MappingsHelper.java

```diff
@@ -0,0 +1,226 @@
+    private static final int SRC_OFFICIAL = 0;
+    private static final int DEST_OFFICIAL = 0;
+
+    private static final MemoryMappingTree officialIntermediaryNamed = new MemoryMappingTree();
+    private static final int SRC_INTERMEDIARY = 0;
+    private static final int DEST_INTERMEDIARY = 0;
+    private static final int SRC_NAMED = 1;
+    private static final int DEST_NAMED = 1;
+
+    private static final boolean IS_DEV_ENV = FabricLoader.getInstance().isDevelopmentEnvironment();
+
+    private static final HttpClient httpClient = HttpClient.newHttpClient();
+
+    public static void initMappings(String version) {
+        HttpRequest versionsRequest = HttpRequest.newBuilder()
+            .uri(URI.create("https://piston-meta.mojang.com/mc/game/version_manifest_v2.json"))
+            .GET()
+            .timeout(Duration.ofSeconds(5))
+            .build();
+        httpClient.sendAsync(versionsRequest, HttpResponse.BodyHandlers.ofString())
+            .thenApply(HttpResponse::body)
+            .thenAccept(versionsBody -> {
+                JsonObject versionsJson = JsonParser.parseString(versionsBody).getAsJsonObject();
+                String versionUrl = versionsJson.getAsJsonArray("versions").asList().stream()
+                    .map(JsonElement::getAsJsonObject)
+                    .filter(v -> v.get("id").getAsString().equals(version))
+                    .map(v -> v.get("url").getAsString())
+                    .findAny().orElseThrow();
+
+                HttpRequest versionRequest = HttpRequest.newBuilder()
+                    .uri(URI.create(versionUrl))
+                    .GET()
+                    .timeout(Duration.ofSeconds(5))
+                    .build();
+                httpClient.sendAsync(versionRequest, HttpResponse.BodyHandlers.ofString())
+                    .thenApply(HttpResponse::body)
+                    .thenAccept(versionBody -> {
+                        JsonObject versionJson = JsonParser.parseString(versionBody).getAsJsonObject();
+                        String mappingsUrl = versionJson
+                            .getAsJsonObject("downloads")
+                            .getAsJsonObject("client_mappings")
+                            .get("url").getAsString();
+
+                        HttpRequest mappingsRequest = HttpRequest.newBuilder()
+                            .uri(URI.create(mappingsUrl))
+                            .GET()
+                            .timeout(Duration.ofSeconds(5))
+                            .build();
+                        httpClient.sendAsync(mappingsRequest, HttpResponse.BodyHandlers.ofString())
+                            .thenApply(HttpResponse::body)
+                            .thenApply(StringReader::new)
+                            .thenAccept(reader -> {
+                                try {
+                                    MappingReader.read(reader, MappingFormat.PROGUARD_FILE, mojmapOfficial);
```

Comment:
Is `MemoryMappingTree` thread safe? It's quite unlikely that the player manages to log into a world before this finishes, but if they do and run the command this will cause a race condition. Also in the future we might decide to use this class elsewhere that gets called earlier. The various access methods below should block while this request is still finishing (or until the request fails, if the case may be).

The mappings should probably be downloaded to a local file which could be used as a fallback in case the user later runs the game without an internet connection. Also this needs to fail gracefully if the user runs the game offline.